### PR TITLE
#143412449 Add JWT token upon successful login

### PR DESCRIPTION
**Pull request by LWanjiru**
#### What does this PR do?
- Add a JWT token which is sent to the user upon successful login.

 #### Description of Task to be completed?
- Users once authenticated via the login api. A JWT should be returned, which is to be required on subsequent call to all other endpoints that require authentication.

 #### How should this be manually tested?
- Specify the endpoint on Postman using the POST method
- Enter the email and password of an existing user in the `body` tab making sure to select the `x-www-form-urlencoded` option, in the provided {Key, Value} format.
**example**
```
email, this@isanexample.com
password, som3pa55word
```
- Click on `Send`

 #### What are the relevant pivotal tracker stories?
143412449

 #### Screenshots (if appropriate)
##### Registered user
screenshot 2017-05-01 17 22 05

##### Unregistered user
screenshot 2017-05-01 17 21 04

##### Wrong email format
screenshot 2017-05-01 17 20 40

##### Empty fields
screenshot 2017-05-01 19 20 49

### CODE REVIEW

**Review by muchai-mercy — APPROVED** [review]

**Review by Ewanjiru — APPROVED** [review]

File: app/controllers/users.js [hunk 1]
```diff
@@ -152,37 +140,61 @@ exports.addDonation = function(req, res) {
...
-};
\ No newline at end of file
+};
+
+ // Attach token to user credentials after authentication
+exports.loginWithEmail = (req, res) => {
+  // Check if the email and passord fields are empty
+  if (req.body.email === '' || req.body.password === '') {
+    return res.status(401).json({ message: 'The email and password fields cannot be empty' });
+  } else if (!(/^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/.test(req.body.email))) {
```
Comment (gngchrs) [inline app/controllers/users.js]: define regex string in a variable
   
    var emailRegex = /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/
     emailRegex.test('email')

File: app/controllers/users.js [hunk 2]
```diff
@@ -1,65 +1,59 @@
...
- * Auth callback
- */
-exports.authCallback = function(req, res, next) {
+// Module dependencies.
+const mongoose = require('mongoose'),
+
+  User = mongoose.model('User'),
+  avatars = require('./avatars').all(),
+  jwt = require('jsonwebtoken'),
+  secretKey = process.env.SECRET_KEY;
```
Comment (gngchrs) [inline app/controllers/users.js]: what if secret is empty/not set?

Comment (gngchrs) [reply app/controllers/users.js]: give a defaultt value...  

       process.env.SECRET_KEY | "default value"

File: app/controllers/users.js [hunk 3]
```diff
@@ -152,37 +140,61 @@ exports.addDonation = function(req, res) {
...
+        return res.status(401).json({
+          message: 'Invalid password'
+        });
+      }
+      // Generate and assign token to authenticated user
+      const token = jwt.sign(user._id, secretKey, {
+        expiresIn: '24h'
+      });
+      // Send token
+      user.password = null;
```
Comment (gngchrs) [inline app/controllers/users.js]: minor: remove

File: app/controllers/users.js [hunk 4]
```diff
@@ -152,37 +140,61 @@ exports.addDonation = function(req, res) {
...
+      res.status(200).json(Object.assign({}, user._id, user.name, user.email, { token }));
```
Comment (gngchrs) [inline app/controllers/users.js]: let object = Object.assign({}, ....)

File: app/controllers/users.js (hunk 2 above)
Comment (LWanjiru) [reply app/controllers/users.js]: Resolved.

File: app/controllers/users.js (hunk 1 above)
Comment (LWanjiru) [reply app/controllers/users.js]: Resolved.

**Pull request closed without merging**